ogr/ogrsf_frmts/adbc/ogradbcdataset.cpp

```diff
@@ -0,0 +1,512 @@
+        return false;
+    }
+
+    std::string osLayerName = "RESULTSET";
+    std::string osSQL;
+    const char *pszSQL = CSLFetchNameValue(poOpenInfo->papszOpenOptions, "SQL");
+    bool bIsParquetLayer = false;
+    if (bIsParquet)
+    {
+        m_osParquetFilename = pszFilename;
+        osLayerName = CPLGetBasename(pszFilename);
+        if (osLayerName == "*")
+            osLayerName = CPLGetBasename(CPLGetDirname(pszFilename));
+        if (!pszSQL)
+        {
+            osSQL = CPLSPrintf(
+                "SELECT * FROM '%s'",
+                CPLString(pszFilename).replaceAll("'", "''").c_str());
+            pszSQL = osSQL.c_str();
+            bIsParquetLayer = true;
+        }
+    }
+
+    if (pszSQL)
+    {
+        if (pszSQL[0])
+        {
+            auto poLayer = CreateLayer(pszSQL, osLayerName.c_str());
+            if (!poLayer)
+                return false;
+            poLayer->m_bIsParquetLayer = bIsParquetLayer;
+            m_apoLayers.emplace_back(std::move(poLayer));
+        }
+    }
+    else if (bIsDuckDB || bIsSQLite3)
+    {
+        auto poLayerList = CreateLayer(
+            "SELECT name FROM sqlite_master WHERE type IN ('table', 'view')",
+            "LAYERLIST");
+        if (!poLayerList || poLayerList->GetLayerDefn()->GetFieldCount() != 1)
+        {
+            return false;
+        }
+
+        for (const auto &poFeature : poLayerList.get())
+        {
+            const char *pszLayerName = poFeature->GetFieldAsString(0);
+            if (bIsSQLite3 && EQUAL(pszLayerName, "SpatialIndex"))
+                continue;
+            const std::string osStatement = CPLSPrintf(
+                "SELECT * FROM \"%s\"",
+                CPLString(pszLayerName).replaceAll("\"", "\"\"").c_str());
+            CPLTurnFailureIntoWarning(true);
+            auto poLayer = CreateLayer(osStatement.c_str(), pszLayerName);
+            CPLTurnFailureIntoWarning(false);
+            if (poLayer)
+            {
+                m_apoLayers.emplace_back(std::move(poLayer));
+            }
+        }
+    }
+    else if (bIsPostgreSQL)
+    {
+        auto poLayerList = CreateLayer(
+            "SELECT n.nspname, c.relname FROM pg_class c "
+            "JOIN pg_namespace n ON c.relnamespace = n.oid "
+            "AND c.relkind in ('r','v','m','f') "
+            "AND n.nspname NOT IN ('pg_catalog', 'information_schema') "
+            "ORDER BY c.oid",
+            "LAYERLIST");
+        if (!poLayerList || poLayerList->GetLayerDefn()->GetFieldCount() != 2)
+        {
+            return false;
+        }
+
+        for (const auto &poFeature : poLayerList.get())
+        {
+            const char *pszNamespace = poFeature->GetFieldAsString(0);
+            const char *pszTableName = poFeature->GetFieldAsString(1);
+            const std::string osStatement = CPLSPrintf(
+                "SELECT * FROM \"%s\".\"%s\"",
+                CPLString(pszNamespace).replaceAll("\"", "\"\"").c_str(),
```

Comment:
If we have an SQL quoting function available somewhere this is when it could be of use.

Reply:
added :
```
/** Replace all occurences of ch by it repeated twice.
 * Typically used for SQL string literal or identifier escaping.
 */
std::string CPL_DLL OGRDuplicateCharacter(const std::string &osStr, char ch);
```